cpp/src/arrow/compute/kernels/hash_aggregate_test.cc

```diff
@@ -2460,6 +2461,294 @@ TEST(GroupBy, Distinct) {
+        if (type->name() == "date64") {
+          EXPECT_THAT(col->GetScalar(0),
+                      ResultWith(AnyOfJSON(type, R"([86400000, 259200000])")));
+          EXPECT_THAT(col->GetScalar(1), ResultWith(AnyOfJSON(type, R"([0])")));
+          EXPECT_THAT(col->GetScalar(2), ResultWith(AnyOfJSON(type, R"([null])")));
+          EXPECT_THAT(col->GetScalar(3),
+                      ResultWith(AnyOfJSON(type, R"([259200000, 432000000])")));
+          EXPECT_THAT(col->GetScalar(4),
+                      ResultWith(AnyOfJSON(type, R"([345600000, 86400000])")));
+        } else {
+          EXPECT_THAT(col->GetScalar(0), ResultWith(AnyOfJSON(type, R"([1, 3])")));
+          EXPECT_THAT(col->GetScalar(1), ResultWith(AnyOfJSON(type, R"([0])")));
+          EXPECT_THAT(col->GetScalar(2), ResultWith(AnyOfJSON(type, R"([null])")));
+          EXPECT_THAT(col->GetScalar(3), ResultWith(AnyOfJSON(type, R"([3, 5])")));
+          EXPECT_THAT(col->GetScalar(4), ResultWith(AnyOfJSON(type, R"([4, 1])")));
+        }
+      }
+    }
+  }
+}
+
+TEST(GroupBy, OneBinaryTypes) {
+  for (bool use_exec_plan : {true, false}) {
+    for (bool use_threads : {true, false}) {
+      for (const auto& type : BaseBinaryTypes()) {
+        SCOPED_TRACE(use_threads ? "parallel/merged" : "serial");
+
+        auto table = TableFromJSON(schema({
+                                       field("argument0", type),
+                                       field("key", int64()),
+                                   }),
+                                   {R"([
+    [null,   1],
+    ["aaaa", 1]
+])",
+                                    R"([
+    ["babcd",2],
+    [null,   3],
+    ["2",    null],
+    ["d",    1],
+    ["bc",   2]
+])",
+                                    R"([
+    ["bcd", 2],
+    ["123", null],
+    [null,  3]
+])"});
+
+        ASSERT_OK_AND_ASSIGN(
+            Datum aggregated_and_grouped,
+            GroupByTest({table->GetColumnByName("argument0")},
+                        {table->GetColumnByName("key")}, {{"hash_one", nullptr}},
+                        use_threads, use_exec_plan));
+        ValidateOutput(aggregated_and_grouped);
+        SortBy({"key_0"}, &aggregated_and_grouped);
+
+        const auto& struct_arr = aggregated_and_grouped.array_as<StructArray>();
+        //  Check the key column
+        AssertDatumsEqual(ArrayFromJSON(int64(), R"([1, 2, 3, null])"),
+                          struct_arr->field(struct_arr->num_fields() - 1));
+
+        const auto& col = struct_arr->field(0);
+        EXPECT_THAT(col->GetScalar(0), ResultWith(AnyOfJSON(type, R"(["aaaa", "d"])")));
+        EXPECT_THAT(col->GetScalar(1),
+                    ResultWith(AnyOfJSON(type, R"(["bcd", "bc", "babcd"])")));
+        EXPECT_THAT(col->GetScalar(2), ResultWith(AnyOfJSON(type, R"([null])")));
+        EXPECT_THAT(col->GetScalar(3), ResultWith(AnyOfJSON(type, R"(["2", "123"])")));
+      }
+    }
+  }
+}
+
+TEST(GroupBy, OneScalar) {
```

Comment:
Is this the correct way to write this?

Reply:
Hmm, this needs to be something like `ValueDescr::Scalar(type)`. See for instance CountScalar: https://github.com/apache/arrow/blob/7236f48d7c534802ebd84daa709aeaba070d6780/cpp/src/arrow/compute/kernels/hash_aggregate_test.cc#L846-L884

Reply:
For this input
```cpp
input.batches = {
      ExecBatchFromJSON({ValueDescr::Scalar(int32()), int64()},
                        R"([[-1, 1], [-1, 1], [7, 1], [23, 1]])"),
      ExecBatchFromJSON({ValueDescr::Scalar(int32()), int64()},
                        R"([[-9, 1], [null, 1], [null, 2], [null, 3]])"),
      ExecBatchFromJSON({int32(), int64()}, R"([[22, 1], [3, 2], [4, 3]])"),
  };
  input.schema = schema({field("argument", int32()), field("key", int64())});
```
this:
```cpp
const auto& struct_arr = actual.array_as<StructArray>();
ARROW_LOG(WARNING) << struct_arr->ToString();
```
has the following output:
```
-- is_valid: all not null
-- child 0 type: int32
  [
    -1,
    -9,
    -9
  ]
-- child 1 type: int64
 [
    1,
    2,
    3
 ]
```

And I can't understand what's going on. Also as in the `CountScalar` example above, there is no `ValueDescr::Scalar()` used in the third call of `ExecBatchFromJSON()`, what is the reasoning behind this?

Reply:
ExecBatch, for each column, can either hold an array (as we've seen so far) or a scalar. A scalar is used to compress the representation when all values for a row in a batch are the same. So first off, the first value in each row above needs to be the exact same value:

```
input.batches = {
      ExecBatchFromJSON({ValueDescr::Scalar(int32()), int64()},
                        R"([[-1, 1], [-1, 1], [-1, 1], [-1, 1]])"),
```

however just because one batch has a scalar doesn't mean that all batches do, that's why the third batch is different above

Reply:
Using `ValueDescr::Scalar()` in all three `ExecBatchFromJSON()` calls produces the same outputs are above too.